### DISCUSSION TITLE
Active-active, eventually consistent replication for JetStream?

### DISCUSSION THREAD

**Discussion by lewisl9029** [nats-io/nats-server#3866] [Q&A]
Hi NATS team! :)

The product I'm building (https://reflame.app) has a bunch of use cases that require global replication and are very sensitive to local write latency, so I've built them on top of a bunch of different tech that allows for "fast" local writes (i.e. no need to wait for ~300ms+ for global consensus) and eventually consistent global replication, namely DynamoDB for app state, S3 for binary blobs, KeyDB for caching, etc.

I've been following the development of NATS for a while, and I absolutely love the idea of unifying these use cases under the same tech with NATS JetStream and the KV/Object store abstractions, but all the replication approaches I could find in the docs seems to be either:

1. uni-directional in the case of mirroring: https://docs.nats.io/running-a-nats-service/nats_admin/jetstream_admin/replication

or

2. requires consensus in the case of clustering https://docs.nats.io/running-a-nats-service/configuration/clustering/jetstream_clustering 

If I missed something and this in fact possible, I'd really appreciate some pointers on how to set it up! Cheers!

**Comment by ripienaar** [comment]
You can have streams `ORDERS_LON`, `ORDERS_SFO`, `ORDERS_NYC` with subjects like `orders.nyc.>` and have a central `ORDERS` that sources it all.  You can also source `ORDERS_SFO` into `ORDERS_LON` for example to create a full view in LON thats eventually consistent.

I havnt tried this but you might combine this with cluster-unique mappings so each cluster can publish to the same subject and get it written to nearest available one magically.

**Comment by derekcollison** [comment]
If you only want 2 streams, you can source them from one another as long as the subject spaces to not overlap.

So imagine you have two streams, A and B. Let's say the subject they listen on collectively is `ORDERS.*.>`. The second token will be origin stream.

So stream A will have subjects `ORDERS.A.>` and stream B will have `ORDERS.B.>`.

Stream A will source from stream B from filtered subject `ORDERS.B.>` and stream B will source from stream A on `ORDERS.A.>`.

The last customization could be that when a user connects to a certain node that house stream A, they inherit a subject mapping that maps `ORDERS.>` to `ORDERS.A.>`, and the same style mapping for a node that house stream B.

This way the app simply sends to ORDERS.CREATED.123 and the message is stored in the local stream but then asynchronously stored in the other stream as well.

**Reply by lewisl9029** [reply]
Thanks @derekcollison and @ripienaar for the super quick responses! The approach you both described sounds super promising!

The only thing I'm still curious about is how to layer the KV/Object store abstractions on top of this kind of setup. 

Ideally I want to have a KV and Object store sit on top of the combined streams in each region, and be able to call `put` on a key to store in the local stream immediately with an ack and replicate eventually to other regional streams, and call `get` on a key to receive the latest known value for the key across all the replicated regional streams.

I was looking at the KV walkthrough docs at https://docs.nats.io/nats-concepts/jetstream/key-value-store/kv_walkthrough and couldn't find anything on how to create a kv from a heavily customized jetstream setup like the one we've described here.

**Reply by computinglife** [reply]
> If you only want 2 streams, you can source them from one another as long as the subject spaces to not overlap.
> 

How is this configured ? Docs does not seem to indicate how i can provide an EXTERNAL server + its stream as the source ?

**Reply by ripienaar** [reply]
You can leafnode them together then native replication works. 

Else replicating KV with an external process is quite easy using a watcher.

**Comment by ripienaar** [comment]
KV mirrors and sources are possible but will o let wkrk for your case in next major server reader. Object not today. 

You can use a normal stream in a KV-like manner but for object I don’t think it’s a good fit.
